### PR TITLE
Add Rail Pool

### PR DESCRIPTION
**Pull request by knexator**
This PR introduces a new genre: Rail Pool by Menderbug.

For some example puzzles, check out Menderbug's twitter: https://twitter.com/menderbug/status/1521766920453771265
FR5mubbXwAAD2OL

Rules: Draw a single orthogonal, non-intersecting loop that visits every cell. If a region contains any numbers, those numbers are the set of lengths of all line segments contained fully or partially within that region (for segments contained partially in the region, the number still refers to the total length of the segment). 
In other words, all line segments have to have one of the given lengths, and each given length matches at least one line segment.

It's a very young genre, but I think it shows great promise, and having a specific player for it might help it reach a wider audience.

This is my first time working with pzprjs; I've mostly copy-pasted stuff from several different genres, not always 100% understanding everything. In addition, there are some extra genre-specific problems that I'm not sure how to solve (see below).

TODO:
- [x] Encoding/Decoding (i could use some help from @x-sheep)
- [x] Add rules
- [x] Better wording on error messages
- [x] Japanese translations of everything
- [x] Tests
- [x] Fix input?

This genre's cluing has two particularities I haven't seen together on any other pzprjs genre: each region might contain up to 2*N clues (with N being the size of the region), and each clue can be a number up to S (with S being the board size).

Since the clues are per-region, I originally took [yajilin-regions](https://puzz.link/p?yajilin-regions) as a base. However, having all the clues cramped into a single cell wont visually work in the general case, due to the first particularity. In addition, Menderbug himself prefers being able to put the clues anywhere (for aesthetics). Switching to tapa-style clues allows placing up to 4*N clues per region, which solves the first problem.

However, multiple clues per region make it hard to input big numbers: both "1, 2" and "12" might be valid inputs. To solve this, clues are artifically limited to be between 1 & 9. For real-world puzzles I don't think it's a problem, but still seems like an inelegant solution.

### CODE REVIEW

**Comment by x-sheep** [comment]
Hi! Thanks for your interest in contributing 👀 I'll hold off on the code review until it's out of WIP.

A few things to note from your post:
* You don't need to have everything translated to Japanese, but you do need to have all FailCodes available in the Japanese language file (or the test will fail). It's okay to duplicate the English strings into the Japanese language file. Our [translation tool](https://hosted.weblate.org/projects/pzprjs/) will automatically mark strings that are the same in both languages, so they can be easily located and fixed.
* The Lohkous implementation is also limited to numbers up to 9. As long as it's supported by the URL encode, I don't think it needs addressing in the interface prior to merging. It can always be added in later.

**Comment by knexator** [comment]
Thanks to you for maintaining this repo!

I hadn't noticed that Lohkous has the same clues, it could have saved me quite a bit of time 😓
Both Lohkous and Rail Loop allow arbitrarily many clues per cell, with clues being arbitrarily big. However, Lohkous encoding is limited to arbitrarily many numbers in the 0-9 range. I think this could be solved by using a-k as single digits signaling that the number still goes on (so 5b78abc4 would decode to 5, 27, 8, 1234). I see two options:
 - Reuse Lohkous's encoding and decoding (maybe moving them to a common file?)
 - Use the other future-proof encoding. In this case, would there be a way to include it in Lohkous without breaking existing puzzles?
 Which one would you recommend, @x-sheep?
 In any case, I will change the current input style to the one Lohkous for consistency

**Comment by x-sheep** [comment]
I think having a Lohkous puzzle with a number above 9 are even less likely than a Rail Pool puzzle with a number above 9. It's safest to leave the Lohkous encoding as-is.

**Comment by knexator** [comment]
Everything ready for review!

File: src-ui/index.html
Comment (x-sheep) [inline src-ui/index.html]: Can you revert this file, then add the Rail Pool link again? The only formatting that's needed is done by running `npm run format`.

File: src/pzpr/variety.js
Comment (x-sheep) [inline src/pzpr/variety.js]: I'd like to keep everything in this object in alphabetical order.

File: src/variety/railpool.js
Comment (x-sheep) [inline src/variety/railpool.js:589]: If you call `setnoerr()` on every border on the board, it grays out all non-errored lines.

File: src/res/failcode.ja.json
Comment (x-sheep) [inline src/res/failcode.ja.json]: Can you sort the keys in both failcode files in alphabetical order?

File: src/res/failcode.en.json
Comment (x-sheep) [inline src/res/failcode.en.json]: This particular error message is pretty confusing. Is it about the amount of distinct lengths in a region that contains question marks?

File: src/res/failcode.en.json
Comment (x-sheep) [inline src/res/failcode.en.json]: Small grammar thing: All failcodes must end with a full stop.

File: src/variety/railpool.js
Comment (x-sheep) [inline src/variety/railpool.js:551]: When I got this error, I initially wasn't sure which region's numbers were causing it. Could you highlight the entire region for this check, like in `checkIncompatibleSets`?

File: src/variety/railpool.js
Comment (x-sheep) [inline src/variety/railpool.js]: It's more common to use `this.errcolor1` for errors instead of bright red, but I'm not sure if I should change Lohkous to use `red` instead...

File: src/variety/railpool.js
Comment (x-sheep) [inline src/variety/railpool.js]: I also don't know why both setQnums and addOpeQnums do a check for no-ops. It seems okay to remove `addOpeQnums` and replace all references to a call to `setQnums`. Adding the ObjectOperation2 should be done in this function.

**Review by x-sheep — CHANGES_REQUESTED** [review]
I have a list of small changes I'd like to see before merging, but no major rewrites necessary.

File: src/variety/railpool.js
Comment (knexator) [reply src/variety/railpool.js:551]: Done, but note that Lohkous doesn't have this feature.

File: src/variety/railpool.js
Comment (knexator) [reply src/variety/railpool.js]: I'm using `red` to emphasize that it's a problem with that individual number and not the whole cell. Right now all the other correct numbers will be `this.errcolor1` since that's the usual color for a number over an errored cell, so there are two options:
 - `this.errcolor1` for all and `red` for the specific number (as it is right now)
 - gray/black for all and  `this.errcolor1` for the specific number

Whichever you prefer
options

File: src/res/failcode.en.json
Comment (knexator) [reply src/res/failcode.en.json]: Yep - this error happens when no there are no valid values for the question marks. For example, a "2 ? ?" clue can never fit in a region with segments of lengths 2 & 3. I've tried several wordings but nothing convinces me - please suggest some alternate error message.

File: src/variety/railpool.js
Comment (x-sheep) [reply src/variety/railpool.js:551]: Lohkous doesn't need the feature, since the error will only show up if the region is completed, and the highlight will be contained entirely within that region. But for Rail Pool, errored line segments can cover any amount of regions.

File: src/variety/railpool.js
Comment (x-sheep) [reply src/variety/railpool.js]: Keeping the other numbers black looks good to me.

File: src/variety/railpool.js
Comment (x-sheep) [reply src/variety/railpool.js:551]: Ah, I see that the region is now highlighted when a number is unused, but not when a segment is invalid. I'd like it to be the other way around.

Comment (knexator) [reply src/variety/railpool.js:551]: Oops, that makes more sense - fixed now.

File: src/variety/railpool.js
Comment (knexator) [reply src/variety/railpool.js]: That comment doesn't only refer to that line: i don't know what Operations are, what's the `klass`, and what's the goal of `ObjectOperation2`. `addOpeQnums` is only called in line 194; without understanding the code, it doesn't seem equivalent to calling `setQnums`, since the latter will always call the `addOpeQnums` functionality and the former does only so depending on the parameter `isrec` (whose origin & function I ignore). Given that the code is the same in tapa, tapaloop, and lohkous, I'm happy leaving it as-is, even if it would be okay to simplify it.

Comment (x-sheep) [reply src/variety/railpool.js]: Alright, it's fine to not merge the functions, as long as this comment is removed.

Operation subclasses are used for the undo stack, and the call to `addOpeQnums` contains the current value and the new value of qnums. Since having multiple qnums in one cell isn't part of the standard framework, Tapa includes a subclass `ObjectOperation2` specifically for replacing the entire list of numbers. This was then copied to Tapa-Like Loop and Lohkous.

File: src/res/failcode.en.json
Comment (knexator) [reply src/res/failcode.en.json]: At Menderbug's suggestion, I've separated this error into two versions: too many & too few questions marks. The wording in these is much clearer.

**Review by x-sheep — APPROVED** [review]

**Comment by x-sheep** [comment]
Alright, this looks good to me. Since I'd like to not release new genres too closely together, I'm planning to merge this on the 24th (one week after Statue Park).

The only thing this type is missing is an entry in `rules.en.yaml`, but I can also do this if required. Unlike with the failcodes, it's preferred to not touch `rules.ja.yaml` unless there's an actual translation available.

**Pull request merged**
